### PR TITLE
Moved conversation to new interfaces

### PR DESCRIPTION
**Pull request by jswistak**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -10,14 +16,87 @@
...
+
+    print_user_message(conv[-1].content)
+    while "q" not in input(
+            f"{Colors.BOLD_BLACK}Press 'q' to quit or any other key to continue: {Colors.END}"
+    ):
+        # Generate response
+        analysis_conv = prompt.generate_conversation_context(conv, ConversationRolesInternalEnum.ANALYSIS, LLMType.GPT4)
+        analysis_assistant_message = analysis_assistant.generate_response(analysis_conv)
+
+        print(f"{Colors.BOLD_BLACK}Assistant: {Colors.END}", end="")
```
Comment (FilipKolodziejczyk) [inline main.py:55]: There is a dedicated function for that: 'print_assistant_message'

**Review by FilipKolodziejczyk — CHANGES_REQUESTED** [review]

File: conversation.py [hunk 2]
```diff
@@ -1,115 +1,97 @@
...
-        latest_conversation_number += 1
-        with open(number_file_path, "w") as number_file:
-            number_file.write(str(latest_conversation_number))
-
-        print("Conversation saved to", conversation_path)
+    def get_conversation(self) -> List[Message]:
+        """Get the conversation."""
+        return self.conversation
+
+    def _add_to_conversation(self, role: ConversationRolesInternalEnum, content: str) -> None:
```
Comment (FilipKolodziejczyk) [inline conversation.py:39]: Same as below

File: conversation.py [hunk 3]
```diff
@@ -1,115 +1,97 @@
...
+                code = code_snippet[6:]
+                cell_idx = self._execute_python_snippet(code)
+                output.append(self._get_cell_output_stream(cell_idx))
+                if self._check_if_plot_in_output(cell_idx):
+                    output[-1] = "Plot generated, but cannot be interpreted in a text format."
+        if len(output) > 0:
+            code_response = self.format_code_assistant_message(code_response, "\n".join(output))
+
+        self._add_to_conversation(role=ConversationRolesInternalEnum.CODE, content=code_response)
+        print_message(self._get_last_message(), Colors.PURPLE)
```
Comment (FilipKolodziejczyk) [inline conversation.py:80]: Send message code prints code?

**Review by FilipKolodziejczyk — CHANGES_REQUESTED** [review]

File: conversation.py [hunk 4]
```diff
@@ -1,115 +1,97 @@
...
+        self.prompt: IPromptManager = prompt
 
     @staticmethod
-    def extract_message_from_response(response: dict) -> dict:
-        """Extract the message from GPT completion."""
-        try:
-            return response.choices[0]["message"]["content"]
-        except Exception as e:
-            raise Exception("Response format incorrect") from e
+    def format_code_assistant_message(message: str, code_output: str) -> str:
```
[the remark this answers is not in the document]
Comment (jswistak) [reply conversation.py:22]: That is not true, we are using this function to forward the message from the `code` assistant to the `analysis` assistant

[the remark this answers is not in the document]
Comment (FilipKolodziejczyk) [reply conversation.py:22]: Ups. That is true. My bad.

**Review by FilipKolodziejczyk — APPROVED** [review]

**Pull request merged**